Fix battery discharge showing on ABUPS Fuel Fired Power Generation

JasonGlazer opened the pull request:
Fixes #4419

JasonGlazer committed:
Removes the use of "ElectricityProduced", "COGENERATION" meter from two battery storage related output variables.

EnergyArchmage commented:
I do not think this correct.  The idea here is that battery storage introduces a loss of produced energy because the round trip through storage isn't 100% efficient.  This change will do away with that accounting of storage losses and I think that is a mistake.  

For this issue you could just change from "Cogeneration" to "Photovoltaics"  but the problem has always been that the battery could be used with either pv or fuel-fired production or a wind turbine.  
While it is true that this metering is showing up in an odd place, that is only because there isn't a better meter nor a better place in the in ABUPS to put this.   I suppose it might be possible to detect what sort of equipment is producing the electricity and shift the setup call between Cogeneration, Photovoltaics, or WindTurbine depending on what is likely to be the source.   

Or you could just add to the explanation of these output variables in the I/O ref to point out that the battery losses appear on the Cogeneration meter by design because of these issues.  

The full fix, which is a lot of work, would be to add another submeter called "Storage" that is analagous to Photovoltaics, WindTurbine, HTGeothermal, etc and then also another new row to the Electric Loads Satisfied table and associated accumulation to the tabular reporting system.

JasonGlazer committed:
Adding storage impact on energy consumption

JasonGlazer committed:
Changed approach to leave change of storage on cogen meter but using "Electricity" instead of "ElectricityProduced" which gets it off the Electric Loads Satisfied table.

JasonGlazer commented:
@EnergyArchmage @mjwitte I changed the approach to match what I think was the intent that Brent suggested. The ABUPS table still looks a little weird using the defect file from:

https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/4000s/4419

and is shown below:

![capture](https://cloud.githubusercontent.com/assets/2582150/10760837/8f2c63e6-7c8c-11e5-874c-1903466ce41e.JPG)

mjwitte commented on src/EnergyPlus/ManageElectricPower.cc:
```diff
@@ -1339,10 +1339,12 @@ namespace ManageElectricPower {
 				SetupOutputVariable( "Electric Storage Discharge Power [W]", ElecStorage( StorNum ).DrawnPower, "System", "Average", ElecStorage( StorNum ).Name );
-				SetupOutputVariable( "Electric Storage Discharge Energy [J]", ElecStorage( StorNum ).DrawnEnergy, "System", "Sum", ElecStorage( StorNum ).Name, _, "ElectricityProduced", "COGENERATION", _, "Plant" );
+				SetupOutputVariable( "Electric Storage Discharge Energy [J]", ElecStorage( StorNum ).DrawnEnergy, "System", "Sum", ElecStorage( StorNum ).Name);
+				ElecStorage( StorNum ).DecrementDrawnEnergy = -1.0 * ElecStorage( StorNum ).DrawnEnergy;
```
@JasonGlazer I believe this line is misplaced.  It needs to be in the same place that .DrawnEnergy is being set.

JasonGlazer replied:
Your right. Thanks.

EnergyArchmage commented:
I think all the code related to DecrementedEnergyStored could be cleaned out and replaced with similiar code for DecrementedEnergyDrawn. 

Array1.Project.MArray.hh may need code analgous to this but I do not understand this part of the code.
    inline MA1c_double DecrementedEnergyStored() const { return ma( &T::DecrementedEnergyStored ); }
    inline MA1_double DecrementedEnergyStored() { return ma( &T::DecrementedEnergyStored ); }

The negative value that looks odd will occur when the state of charge drops over the simulation -- when the battery charging is not keeping up and the battery is full at the start and empty at the end.  This is confusing when the information does not appear along with the Net Decrease in On-Site Storage in the Electric Loads Satisfied table.  Perhaps you could do a one-time adjustment to ABUPS data at the end of the run period to adjust the end use to account for Net Decrease in On-Site Storage.  It may double count if not done carefully.

JasonGlazer committed:
Fixed location of DecrementDrawnEnergy calcs

JasonGlazer commented:
I just fixed the location of the DecrementDrawnEnergy calcs and the results are really:

![capture1](https://cloud.githubusercontent.com/assets/2582150/10762429/06dc8dba-7c94-11e5-9627-d2bf1affc36f.JPG)
![capture2](https://cloud.githubusercontent.com/assets/2582150/10762432/0937569e-7c94-11e5-8723-af1052a2f148.JPG)

JasonGlazer commented:
I am not sure I understand the inline code changes either but it seems to work without them.

mjwitte commented:
A negative value in the main ABUPS table doesn't make sense here.  The storage isn't creating energy.

EnergyArchmage commented:
Okay, I checked this branch out, built it and ran the example files ShopWithPVandStorage and ShopWithPVandBattery.  ShopWithPVandBattery and the defect file use the newer kinetic lead acid model and they both come out with strange negatives there.  ShopWithPVandStorage works as expected. 

It looks like the kinetic battery model got  implemented with a different sign convention than the simple storage bucket.  So the metering is/was all backwards for that one and this is a serious issue of incorrect metering in addition to the issue of how reporting is organized.

EnergyArchmage committed:
Change sign convention in ElectricLoadCenter:Storage:Battery

flip sign convection in kinetic battery model to match simple storage
model and what is expected for metering losses.

JasonGlazer commented:
Thanks Brent. Here are the tables from the ABUPS report:

![capture3](https://cloud.githubusercontent.com/assets/2582150/10787751/22c0e532-7d41-11e5-87bb-5a34aa8fc713.JPG)
![capture4](https://cloud.githubusercontent.com/assets/2582150/10787754/25228f6a-7d41-11e5-9b8e-3ad9dff46ae9.JPG)

The sign for generators on the end use table looks more reasonable.

mjwitte commented:
But the magnitude seems terribly high - more power lost in the storage than produced by the PV?  How can that be if the only source that can charge the battery is the on-site PV generation?

EnergyArchmage commented:
Yeah, that is not good.  ShopWithPVandBattery does not have the same problem.  Can you refresh the defect file you are using?

JasonGlazer commented:
I am not sure what you mean by refresh. The defect files are at:

https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/4000s/4419

EnergyArchmage committed:
revert sign change on kinetic battery discharge power

also added end use subcat names for inverter and transfer

EnergyArchmage commented:
oh sorry, I see a v8.4 file now, _I_ needed to refresh.  I see now that the discharge power didn't need its sign flipped and reverted that change here https://github.com/NREL/EnergyPlus/pull/5263#.   There are other issues with the kinetic battery model that would need to get fixed before the results will be good.  I think you want to restrict this to just the original tabular reporting issue and suggest to use example file ShopWithPVandStorage.idf as the defect file.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 93653546-FuelGenShowingInABUPSwhenPV

mjwitte commented:
Merged in develop to get clean test results.

EnergyArchmage commented:
Note that the test failures here are expected.  End use subcategories were added for batteries, inverters, and transformers and they show up as string diffs in html reports and changes in the table sizes.

mjwitte commented:
@EnergyArchmage But at least one case has html file changes that I can't explain.  Look at eplustbl.htm.absdiff.htm [here](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2015-10/621b8629c56175639755f366ce858e44b175086c-bbd7b7528e5429ffa7439b5ba7b0e101fb387610/ShopWithSimplePVT/Win64-Windows-7-VisualStudio-12/).  Why are they small changes in time setpoint not met and time not comfortable?

EnergyArchmage commented:
The reason is that the transformer is pulling the on-site power generation from  "Cogeneration:ElectricityProduced", which has changed, and the transformer has thermal losses to the zone.  This shows a very good reason why these losses do need to be moved over to the be end uses. 

But...  I noticed that transformer losses also get put on the the fuel-fired generator (nastly little feedback loop there I suspect). 
                SetupOutputVariable( "Transformer Cogeneration Electric Loss Energy [J]", Transformer( TransfNum ).ElecProducedCoGen, "System", "Sum", Transformer( TransfNum ).Name, _, "ElectricityProduced", "COGENERATION", "Transformer", 

Anyway so the transformer losses also need to be moved as they should follow the battery's pattern.  I will push up some code for that shortly.

mjwitte commented:
@EnergyArchmage Be aware of #4622 which added the transformers to an end-use so that the total electricity in the ABUPS report would match the sum of the rows.

EnergyArchmage commented:
AH shoot, it isn't the transformer, there isn't even one.  It is a radically different amount of charging and discharging in the battery.  This whole approach of metering as an end use is not working well at all.  This has to be rejected entirely.   It is circular, the amount of power left over for storage cannot depend on the amount of power that got stored.  

(It may not be such a problem for the utility-feed version of the transformer in PR 4622). 

Changing the end uses affects the Facility Total Electric Demand Power which is a major input into the controlling of charging and discharging.   Here is plot from ShopWithSimplePVT comparing this variable between this PR and develop.

![facility elec demand](https://cloud.githubusercontent.com/assets/8754769/10855378/94f305ae-7f17-11e5-837d-d7e1733e6668.png)

Sorry, I go back to my original suggestion for the proper fix  "to add another submeter called "Storage" that is analagous to Photovoltaics, WindTurbine, HTGeothermal, etc and then also another new row to the Electric Loads Satisfied table and associated accumulation to the tabular reporting system."

mjwitte commented:
Closing this, replaced by #5300
